lib/src/method_channel.dart

```diff
@@ -37,38 +54,63 @@ class ShowMenuArgs {
   Map<String, dynamic> toJson() {
     return {
       'devicePixelRatio': devicePixelRatio,
-      'position': [position.dx, position.dy],
+      'position': <double>[position.dx, position.dy],
       'items': items.map((e) => e.toJson()).toList(),
     };
   }
 }
 
-const _channel = MethodChannel('native_context_menu');
+final _channel = const MethodChannel(_kChannelName)
+  ..setMethodCallHandler(
+    (call) async {
+      switch (call.method) {
+        case _kOnItemSelected:
+          {
+            _contextMenuCompleter.complete(call.arguments);
+            break;
+          }
+        case _kOnMenuDismissed:
+          {
+            _contextMenuCompleter.complete(null);
+            break;
+          }
+        default:
+          {
+            _contextMenuCompleter.completeError(
+              Exception('$_kChannelName: Invalid method call received.'),
+            );
+          }
+      }
+    },
+  );
+
+Completer<int?> _contextMenuCompleter = Completer<int?>();
+
 int _menuItemId = 0;
 
 Future<MenuItem?> showContextMenu(ShowMenuArgs args) async {
   final menu = _buildMenu(args.items);
   _menuItemId = 0;
 
-  final id = await _channel.invokeMethod('showMenu', args.toJson());
+  _channel.invokeMethod(_kShowMenu, args.toJson());
 
-  if (id != -1) {
-    return menu[id];
-  }
+  final id = await _contextMenuCompleter.future;
+  _contextMenuCompleter = Completer<int?>();
+
+  return menu[id];
```

Comment:
When `id` is `null`, it automatically returns `null`. I feel there is no need for another check.